graph/preprocessor.go

```diff
@@ -0,0 +1,211 @@
+	// Iterating in reverse to easily and efficiently handle hierarchy. The later
+	// declared the higher in the hierarchy of alias definitions.
+	for i := len(preTask.AliasSrc) - 1; i >= 0; i-- {
+		aliasURI := *preTask.AliasSrc[i]
+		if strings.HasPrefix(aliasURI, "https://") || strings.HasPrefix(aliasURI, "http://") { // Rewrite in nice case insensitive regex
+			if err := addAliasFromRemote(preTask, aliasURI); err != nil {
+				return err
+			}
+		} else {
+			if err := addAliasFromFile(preTask, aliasURI); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+/* Fetches and Parses out remote alias files and adds their content
+to the passed in PreTask. Note alias definitions already in preTask
+will not be overwritten. */
+func addAliasFromRemote(preTask *PreTask, url string) error {
+	remoteClient := http.Client{
+		Timeout: time.Second * 2, // Maximum of 2 secs
+	}
+
+	req, err := http.NewRequest(http.MethodGet, url, nil)
+	if err != nil {
+		return err
+	}
+
+	res, getErr := remoteClient.Do(req)
+	if getErr != nil {
+		return getErr
+	}
+
+	data, readErr := ioutil.ReadAll(res.Body)
+	if readErr != nil {
+		return readErr
+	}
+
+	return readAliasFromBytes(data, preTask)
+}
+
+/* Parses out local alias files and adds their content to the passed in
+PreTask. Note alias definitions already in preTask will not be
+overwritten. */
+func addAliasFromFile(preTask *PreTask, fileURI string) error {
+
+	data, fileReadingError := ioutil.ReadFile(fileURI)
+	if fileReadingError != nil {
+		return fileReadingError
+	}
+	return readAliasFromBytes(data, preTask)
+}
+
+/* Parses out alias  definitions from a given bytes array and appends
+them to the PreTask. Note alias definitions already in preTask will
+not be overwritten even if present in the array. */
+func readAliasFromBytes(data []byte, preTask *PreTask) error {
+
+	aliasMap := &map[string]string{}
+
+	if err := yaml.Unmarshal(data, aliasMap); err != nil {
+		return err
+	}
+
+	for key, value := range *aliasMap {
+		if _, ok := preTask.AliasMap[key]; !ok {
+			preTask.AliasMap[key] = value
+		}
+	}
+	return nil
+}
+
+// PreprocessString handles managing alias definitions from a provided string definitions expected to be in JSON format.
+func PreprocessString(preTask *PreTask, str string) (string, error) {
+	//preTask.loadGlobalDefinitions TODO?
+
+	// Load Remote/Local alias definitions
+	if externalDefinitionErr := preTask.loadExternalAlias(); externalDefinitionErr != nil {
+		return "", externalDefinitionErr
+	}
+	//Validate alias definitions
+	if improperFormatErr := preTask.resolveMapAndValidate(); improperFormatErr != nil {
+		return "", improperFormatErr
+	}
+	var out strings.Builder
+	var command strings.Builder
+	ongoingCmd := false
+
+	// Search and Replace
+	for _, char := range str {
+		if ongoingCmd {
+			//Maybe just checking if non alphanumeric, only allow alpha numeric aliases?
+			if strings.Contains(")}/ .,;]&|'~\n\t", string(char)) { // Delineates the end of an alias
```

Comment:
seems you need to check it is not in the allowed alias name charset

Reply:
That... Is actually a much better idea than what I am doing. Thanks Bin!